devel/validate-platforms

```diff
@@ -46,8 +46,10 @@ main() {
         # in addition to other important software.
         echo "[$platform] Determining software versions..."
         docker-run "$platform" bash -c '
-            PS4="\$ "
-            set -x
+            function echo-command {
+                echo "$ $BASH_COMMAND"
+            }
+            trap echo-command DEBUG
```

Comment:
_non-blocking_

I don't understand the reason for this change, even after reading the good commit message:

>  Use trap instead of set -x
>
> This allows errors such as "command not found" to terminate the script.
>
>  I considered BASH_XTRACEFD but that isn't available on bash <4.0, which
> is where the default bash version of macOS is tied to indefinitely.

Why does `set -x` not allow errors such as "command not found" to terminate the script?

(I was also thinking, "Do we _want_ those errors to terminate it? or should we let them pass and get picked up in the diff later?", but realized the edge case there is some command being missing from all platform images and thus producing no diff.)

Reply:
Also, FWIW, a more minimal but equivalent trap would be:

```bash
trap 'echo "$ $BASH_COMMAND"' DEBUG
```

Reply:
Oh, maybe the commit message could be better!

> Why does `set -x` not allow errors such as "command not found" to terminate the script?

Because `set -x` outputs to `stderr` by default, so there's no way to differentiate the outputted commands from actual errors when redirecting `stderr` to the report. This can be changed by setting `BASH_XTRACEFD` but I went against that for reasons in the commit message.

Reply:
> ```shell
> trap 'echo "$ $BASH_COMMAND"' DEBUG
> ```

I was trying to use something like this but couldn't figure out how to get it to work properly while inside the `bash -c '…'` literal string. I just found a workaround though:

```
bash -c $'
    trap \'echo "$ $BASH_COMMAND"\' DEBUG
    …
' >"$report"
```

Reply:
> realized the edge case there is some command being missing from all platform images and thus producing no diff.

Yes, I realized when I tried `treetime --version` and noticed the unusual behavior.

Reply:
_continues to be non-blocking_

> Because `set -x` outputs to `stderr` by default, so there's no way to differentiate the outputted commands from actual errors when redirecting `stderr` to the report. This can be changed by setting `BASH_XTRACEFD` but I went against that for reasons in the commit message.

Oh. Hmm. Another way without using `BASH_XTRACEFD` is to

    set -x
    exec 2>&1

if you want to merge stdout and stderr.

But I think I'm still missing something, because it went from using `set -x` to stderr with:

    docker run … >report 2>&1

to `trap` + `echo` to stdout with:

    docker run … >report

Where those producing different interleavings of lines in `report`? If so, that's probably a Docker issue around how it handles stdio ("unusually", is the answer), or potentially with the `docker run` invocation in this PR.

OH! Nevermind, I finally get that you _want_ to separate stdout and stderr, but want the command tracing to go to stdout. Got it. Sorry for being slow on the uptake here.

But speaking of the `docker run` invocation, I just noticed something there… will comment separately.

> but couldn't figure out how to get it to work properly while inside the `bash -c '…'` literal string. I just found a workaround though:

You probably already are, but if not, beware that `$'…'` strings have other semantic changes as well, such as replacing C-style escapes like `\t` with a literal tab.

The more typical way to do it without `$'…'` is to use different kinds of quoted strings adjacent to each other (as adjacent quoted strings are treated as one; Python does something similar):

```bash
bash -c '
    trap '"'"'echo "$ $BASH_COMMAND"'"'"' DEBUG
    …
' >"$report"
```

This is admittedly very ugly and you have to carefully count/match quotes when reading/writing, but it's idiomatic and readily identified on sight as an area of "here be dragons".

Reply:
(Thanks for being patient with me!)

Reply:
(The `function echo-command` approach also remains just fine. No need for a change here, hence _non-blocking_, I just wanted to understand.)

Reply:
Ok, I'm glad we are on the same page now! Thanks for taking the time to understand. I'll update the description of 7d8e9480ba91af55f0a225718b405f63570a465e with the other alternatives, but keep the actual changes as-is.

Reply:
Reworded in b8135dbd719112f513a5a5747737efd55a7315e0.